### PR TITLE
[Discuss] New docs/examples in preparation for async callbacks

### PR DESCRIPTION
**Pull request by bendk**
This PR is how I think async callbacks could work and why they would be useful.  Any feedback would be greatly appreciated.

The main point is that if we have async callbacks, then foreign application that uses the UniFFIed library can drive things.  We don't need to start up new threads or run a tokio event loop.   This opens up some very interesting possibilities.  Check out the `user-store` example for details here.

I also want to suggest a couple ergonomics changes for callbacks, check out `callback_interfaces.md` for details.

### CODE REVIEW

File: docs/manual/src/udl/callback_interfaces.md [hunk 1]
```diff
@@ -1,107 +1,55 @@
...
-```
+All methods of the Rust trait should return a Result.  This is needed because
+any callback method call can fail unexpected ways that shouldn't result in a
+panic, for example when the foreign code throws an exception we weren't
+expecting.
 
-In Rust:
+If your method can only fail in these unexpected ways, use
+`uniffi::CallbackResult<T>` as your result type.  This is a type alias for
+`Result<T, uniffi::UnexpectedCallbackError>`.
```
Comment (bendk) [inline docs/manual/src/udl/callback_interfaces.md:44]: This is a small change that I think could improve ergonomics.  If a callback interface method isn't expected to raise an error, then we shouldn't need to define an error type for it.

File: docs/manual/src/udl/callback_interfaces.md [hunk 2]
```diff
@@ -1,107 +1,55 @@
...
-
-## 1. Define a Rust trait
-
-This toy example defines a way of Rust accessing a key-value store exposed
-by the host operating system (e.g. the key chain).
+# Defining callback interfaces
 
 ```rust,no_run
-pub trait Keychain: Send + Sync + Debug {
+#[uniffi::callback_interface]
```
Comment (bendk) [inline docs/manual/src/udl/callback_interfaces.md:18]: I think that if we want async we should definitely lean into the macros.  Rust currently doesn't allow async functions in traits, but a macro can convert those to a concrete type to work around it.  We could also support this in UDL, but then you'd need to write some slightly ugly Rust code -- async functions would need to be re-written as sync functions that returned `uniffi::ForeignFuture<T>`.

File: docs/manual/src/udl/callback_interfaces.md [hunk 3]
```diff
@@ -168,3 +153,71 @@ Take care to avoid this by following guidelines:
    references through intermediate objects.
 3. If you need to break the first 2 guidelines, then take steps to manually break the cycles to avoid memory leaks.
    Alternatively, ensure that the number of objects that can ever be created is bounded below some acceptable level.
+
+## Scheduling tasks using the `TaskQueue` callback interface
```
Comment (bendk) [inline docs/manual/src/udl/callback_interfaces.md:157]: This is orthogonal to user-defined callback interfaces, but I think it could be really useful.  I'm thinking we should implement them at more-or-less the same time.

File: examples/user-data-store/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    // Wrapper for a database operation that awaits on DB initialization and loading the crypto
+    // key.  f should be a synchronous function.
+    //
+    // This is the main point of this example.  This wrapper uses async callback interfaces to:
+    //   - Initialize the database and load the crypto key asynchronously
+    //   - Schedule the database operation in a background thread
+    //   - Log any errors
+    //   - Notify the application that the operation completed
```
Comment (bendk) [inline examples/user-data-store/src/lib.rs:98]: This feels like a nice version of [decorators](https://github.com/mozilla/uniffi-rs/pull/1051) to me.

File: examples/user-data-store/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! An example of a user data store that with an async interface that's driven by the executor on
+//! the foreign side of the FFI.  This is achieved by awaiting async callback interface methods.
+//! See the callbacks section of the manual for the technical details.
```
Comment (bendk) [inline examples/user-data-store/src/lib.rs:7]: I expect a bunch of compiler errors if you tried to build this.  The code will definitely need to change, but I think the general spirit is correct.

File: docs/manual/src/udl/callback_interfaces.md [hunk 6]
```diff
@@ -125,36 +81,65 @@ class SwiftKeychain: Keychain {
...
+  pub fn get_password(&self) -> String {
+    UserInfo {
+        username: self.keychain.get("username".into()),
+        password: self.keychain.get("password".into()),
+    }
+  }
 
-Also note, that storing the `Box<dyn Keychain>` in the `Authenticator` required that all implementations
-*must* implement `Send`.
+  pub async fn get_password_async(&self) -> String {
```
Comment (badboy) [inline docs/manual/src/udl/callback_interfaces.md:131]:
```suggestion
  pub async fn get_password_async(&self) -> UserInfo {
```

File: docs/manual/src/udl/callback_interfaces.md (hunk 3 above)
Comment (badboy) [reply docs/manual/src/udl/callback_interfaces.md:157]: Not to say we shouldn't, but from what I read here it might be possible to define and implement this outside of UniFFI.
If so, what benefits does it have to make that part of uniffi directly?

**Review by badboy — COMMENTED** [review]
(note: I already did some typo/other nitpicks review, feel free to ignore this in this early stage of the discussion)

I think the async interfaces overall look good and possible to support.
If we all agree that proc-macros are the future, then I agree that only supporting async trait functions there is fine.
I'm less enthusiastic about the task queue, but not outright blocking. See inline. (fwiw if that would have existed in the past Glean would have probably ended up using it)

File: docs/manual/src/udl/callback_interfaces.md (hunk 3 above)
Comment (bendk) [reply docs/manual/src/udl/callback_interfaces.md:157]: The issue is that it needs a couple of things not supported in UniFFI:
  - Storing/calling an arbitrary Rust closure
  - Generics (it returns the closure return type)

Both of those could be nice features to support in general and once we do, then we don't need to special case this.  However,  I don't see that happening anytime soon.

File: docs/manual/src/udl/callback_interfaces.md [hunk 7]
```diff
@@ -111,9 +59,17 @@ class KotlinKeychain: Keychain {
...
-    override fun put(key: String) {
+    override fun put(key: String, value: String) {
         // … elide the implementation.
     }
+
+    override async fun getAsync(key: String) = withContext(coroutineContext) {
+        get(key)
+    }
+
+    override async fun putAsync(key: String, value: String) = withContext(coroutineContext) {
```
Comment (jhugman) [inline docs/manual/src/udl/callback_interfaces.md:70]: Nit:
```suggestion
    override suspend fun putAsync(key: String, value: String) = withContext(coroutineContext) {
```

Thinking about it however, doesn't this dance with the `coroutineContext` and `suspend` simply disappear now? (Unless you're intent was to show the kotlin running on a different queue to the `TaskQueue` that was passed in at construction?

**Review by jhugman — APPROVED** [review]
This looks very promising/exciting.

Abstracting the various coroutines, dispatch queues, main loops etc into `ThreadQueue` looks a very elegant way to solve the majority of the async use cases we have.

File: docs/manual/src/udl/callback_interfaces.md (hunk 7 above)
Comment (bendk) [reply docs/manual/src/udl/callback_interfaces.md:70]: Good point.  I think you're right, although I'd need to test it to be sure.

File: docs/manual/src/udl/callback_interfaces.md (hunk 3 above)
Comment (mhammond) [reply docs/manual/src/udl/callback_interfaces.md:157]: don't mean to bikeshed too much, but I wonder if something like `TaskExecutor` is a better name here?

File: docs/manual/src/udl/callback_interfaces.md [hunk 8]
```diff
@@ -168,3 +153,71 @@ Take care to avoid this by following guidelines:
...
+val userStore = UserStore(TaskQueue(Dispatchers.IO))
+```
+
+## Async callbacks are driven by foreign executors
+
+Code that awaits an async callback interface method will be woken up by an FFI call from the UniFFI
+foreign bindings code. This allows your Rust code to be driven by the foreign async executor,
+avoiding the need to spawn threads from Rust or run a `tokio` loop.
+
+See the `user-data-store` example for how this could work in action.
```
Comment (mhammond) [inline docs/manual/src/udl/callback_interfaces.md:212]: You intend fleshing that example out with foreign bindings?

File: docs/manual/src/udl/callback_interfaces.md [hunk 9]
```diff
@@ -168,3 +153,71 @@ Take care to avoid this by following guidelines:
...
+
+## Fire and forget callbacks
+
+As mentioned above, UniFFI uses the `ForeignFuture<T>` type to handle async callback interface
+returns.  `ForeignFuture` also implements the `fire()` method, which schedules the future to be run
+by the foreign executor, ignoring the result.  This is good for fire-and-forget style callbacks,
+like logging.
```
Comment (mhammond) [inline docs/manual/src/udl/callback_interfaces.md:219]: ideally that new example would demonstrate that too, so another "See the example" note seems worthwhile

File: examples/callbacks/README.md [hunk 10]
```diff
@@ -0,0 +1,3 @@
+This example is actually several mini-examples that should how callbacks can be used.  The
+emphasis on how they can be used as a port/adapter style system with the embedding application
```
Comment (mhammond) [inline examples/callbacks/README.md:2]: IMO you should replace "port/adaptor style" with a few more words that describes what you mean, as those terms don't appear anywhere else in the repo best I can tell and different people often have different pictures for these generic terms for patterns.

**Review by mhammond — APPROVED** [review]
Wow, this is magic and offers some great capabilities! I'm approving this even though I hope you don't intend landing this PR but instead will roll it up into the implementation?

**Comment by bendk** [comment]
Closing this one, thanks for all the feedback.

Hopefully the next PR I open will contain these docs, with the typos fixed, plus a working implementation.
